Add a plugin mechanism to provide custom locator URL schemes

jandubois opened the pull request:
The template reader will attempt to call `limactl-url-$SCHEME` to rewrite a custom URL as either a local filename, or a URL with a supported scheme like `https`.

This change allows us to experiment with custom schemes, like the one requested in #3930.

Example implementation for the `github:` scheme is at the end. You can use:

```
github:ORG/REPO
github:ORG/REPO/DIR/
github:ORG/REPO/DIR/FILE
github:ORG/REPO/DIR/FILE@BRANCH
```

`FILE` defaults to `lima`, and files without an extension will have `.yaml` appended.

The default branch is determined by the GitHub API if no `BRANCH` is specified.

```console
❯ limactl tmpl copy github:lima-vm/lima/templates/fedora-42 -
minimumLimaVersion: 1.1.0

base:
- template://_images/fedora-42
- template://_default/mounts
```

This implementation has a few differences to the one requested in #3930:
* uses default branch instead of hard-coded `master`
* uses `lima.yaml` instead of `templates/default.yaml` as the default file

I believe this implementation is more versatile, but we can discuss this below. The plugin format makes it easy to experiment.

Here is a minimalistic `my:` scheme to fetch default templates from the Lima repo:

```console
❯ cat ~/bin/limactl-url-my
#!/bin/bash
echo "https://raw.githubusercontent.com/lima-vm/lima/master/templates/$1.yaml"

❯ limactl tmpl copy my:opensuse-leap -
minimumLimaVersion: 1.1.0

base:
- template://_images/opensuse-leap
- template://_default/mounts
```

And here is `limactl-url-github`. Which should become a builtin scheme (implemented in Go) once we agree on the specific semantics:

```bash
#!/bin/bash

if [ $# -ne 1 ]; then
    echo "Usage: $0 [github:]ORG/REPO[/PATH][@BRANCH]" >&2
    exit 1
fi

input="$1"

# Remove optional github: prefix
if [[ "$input" == github:* ]]; then
    input="${input#github:}"
fi

# Check for explicit branch specification with @ at the end
branch=""
if [[ "$input" == *@* ]]; then
    # Split on last @ to get branch
    branch="${input##*@}"
    input="${input%@*}"
fi

# Split the input into components
IFS='/' read -ra PARTS <<< "$input"

if [ ${#PARTS[@]} -lt 2 ]; then
    echo "Error: Input must be at least ORG/REPO" >&2
    exit 1
fi

org="${PARTS[0]}"
repo="${PARTS[1]}"

# Extract path (everything after ORG/REPO)
if [ ${#PARTS[@]} -gt 2 ]; then
    path=""
    for ((i=2; i<${#PARTS[@]}; i++)); do
        if [ -n "$path" ]; then
            path="$path/"
        fi
        path="$path${PARTS[i]}"
    done
else
    path="lima"
fi

# If path ends with /, it's a directory, so append lima
if [[ "$path" == */ ]]; then
    path="${path}lima"
fi

# If the filename (last component) has no extension, add .yaml
filename="${path##*/}"
if [[ "$filename" != *.* ]]; then
    path="$path.yaml"
fi

# Query default branch if no branch was specified
if [ -z "$branch" ]; then
    branch=$(gh repo view "$org/$repo" --json defaultBranchRef --jq '.defaultBranchRef.name' 2>/dev/null)
    if [ $? -ne 0 ] || [ -z "$branch" ]; then
        echo "Error: Failed to get default branch for $org/$repo. Make sure the repository exists and you have access." >&2
        exit 1
    fi
fi

echo "https://raw.githubusercontent.com/$org/$repo/$branch/$path"
```

jandubois commented:
I just realized that the `limactl-url` prefix makes the names overlap with regular plugin names. Not sure if that is an issue or a feature:

```console
❯ limactl url-github lima-vm/lima
https://raw.githubusercontent.com/lima-vm/lima/master/lima.yaml
```

jandubois commented:
I've just tested, and you can use tags or commit ids instead of branch names too. These all work:

```console
❯ limactl tmpl copy github:lima-vm/lima/examples/opensuse@v0.23.1 -
❯ limactl tmpl copy github:lima-vm/lima/examples/opensuse@74e2fda81 -
```

msgilligan commented:
This looks great! 

> a few differences to the one requested in #3930

I did say "something like" in the request to leave room for improvement!

jandubois commented:
I've updated the PR by refactoring the code to run subcommands with the `limactl` directory first on the `PATH` and also added a `limactl template url CUSTOM_URL` command (mostly for integration tests later):

```
❯ l tmpl url github:lima-vm/lima
https://raw.githubusercontent.com/lima-vm/lima/master/lima.yaml
```

AkihiroSuda commented:
What is the remaining task to merge this PR? CI?

jandubois commented:
> What is the remaining task to merge this PR? CI?

I was planning to move to `template:default` style templates (and translate the existing form with a warning to the new format), but also to move the `github:` scheme to be builtin, implemented in Go. I was hoping to get some feedback if the semantics of the bash prototype seem correct to everyone.

Either or both could happen in separate PRs to keep the PR easier to review.

jandubois commented:
> I was planning to move to `template:default` style templates (and translate the existing form with a warning to the new format), but also to move the `github:` scheme to be builtin, implemented in Go. I was hoping to get some feedback if the semantics of the bash prototype seem correct to everyone.

I think it will be best to create separate PRs for further changes, to keep things easy for review.

I've made one more change in this PR: I've appended `/usr/local/libexec/lima` to the end of the `PATH`, so we look for plugins in the same directory we use to store external drivers. This is based on discussion in https://github.com/lima-vm/lima/pull/3744#issuecomment-3250472833.

I've noticed that we also use `LIMA_DRIVERS_PATH` and `LIMA_TEMPLATES_PATH` to define additional search locations for drivers and templates. And we look in `$LIMA_HOME/_templates` for templates as well. I feel like this is getting out of hand and don't want to add similar mechanisms for plugins, at least not yet. Let me know if you think otherwise.

AkihiroSuda commented on cmd/limactl/template.go:
```diff
@@ -95,6 +96,25 @@ func fillDefaults(ctx context.Context, tmpl *limatmpl.Template) error {
 	return err
 }
 
+func newTemplateURLCommand() *cobra.Command {
```
Needs docs and tests

AkihiroSuda replied:
Docs can be added under https://deploy-preview-3972--lima-vm.netlify.app/docs/config/plugin after merging:
- #3972

jandubois replied:
This PR is also modifying the same logic that #3973 is currently changing.

And tests should be implemented using BATS from #3969.

I'm turning this back unto "Draft" until these 3 PRs have landed because it will likely need serious refactoring by then.

jandubois commented:
I just created another URL scheme for testing: `instance:` to reference the template of an already created instance:

```bash
#!/bin/bash
echo "${LIMA_HOME:-$HOME/.lima}/$1/lima.yaml"
```

With this you don't actually need `limactl clone` because you can just write:

```console
❯ limactl create -y --name clone instance:default
```

And of course it works all the usual places: `limactl tmpl validate instance:clone` or `limactl tmpl copy instance:default -` etc.

Not sure though if it is useful enough to be builtin.

AkihiroSuda commented:
> instance:

This is quite confusing; `limactl start instance:NAME` should just mean starting an instance named `NAME` (i.e., `limactl start NAME`)

jandubois commented:
> > instance:
> 
> This is quite confusing; `limactl start instance:NAME` should just mean starting an instance named `NAME` (i.e., `limactl start NAME`)

Yeah, let's not include it in Lima. I think the only scheme we want to have builtin is `github`.

jandubois commented:
> ```
> github:ORG/REPO
> github:ORG/REPO/DIR/
> github:ORG/REPO/DIR/FILE
> github:ORG/REPO/DIR/FILE@BRANCH
> ```

Extension ideas:

* `github:ORG` is the same as `github:ORG/lima.yaml` or `github:ORG/ORG` if `lima.yaml` repo doesn't exist
* `github:ORG//` is the same as `github:ORG`, but you can now append a DIR, FILE, and BRANCH: `github:mylongprojectname//templates/lima.yaml`

jandubois commented:
I think this PR is ready for review.

Still missing, but could be done in several separate PRs:
* Implement builtin `github:` scheme in Go
* Switch to `template:` scheme, but keep `template://` working for backwards compatibility
* Add BATS tests for plugins in general and url-* plugins in particular
* Explain url-* plugins in the docs

AkihiroSuda commented on pkg/limatmpl/locator.go:
```diff
@@ -285,3 +295,55 @@ func InstNameFromYAMLPath(yamlPath string) (string, error) {
+		}
+		// Fix malformed "template:" URLs.
+		newLocator := "template:" + path.Join(u.Host, u.Path)
+		logrus.Warnf("Template locator %q should be written %q", locator, newLocator)
```
```suggestion
		logrus.Warnf("Template locator %q should be written %q since Lima v2.0", locator, newLocator)
```

AkihiroSuda replied:
This needs to be updated too
https://github.com/lima-vm/lima/blob/94ba0e1c742bad6cd76452b07f2161bc9d720971/cmd/limactl/completion.go#L27-L49

jandubois replied:
I updated `cmd/`, `hack/`, and `README.md`.

I've not yet updated `templates/` and `website/`. I assume we only want to update these close to the release, or should I update everything right away?

jandubois replied:
> I've not yet updated `templates/`

And I had to revert `hack/test-upgrades.sh` to keep the command compatible with the old Lima version.

jandubois commented on README.md:
````diff
@@ -39,14 +39,14 @@ lima nerdctl run --rm hello-world
 To run containers with Docker:
 ```bash
-limactl start template://docker
+limactl start template:docker
````
I kind of disagree. The website should document the released version, but the `README.md` should correspond to the state of the branch.

Otherwise we could also not close issues until their PRs have been included in a new release, which would make the whole workflow really awkward.

I'll revert for now, but after 2.0 we should rethink our workflow, and also figure out how to have a version of the docs for the last release, and a different version for the head of `master`.

AkihiroSuda replied:
Regardless to our intention, people will just refer to the README in the default branch and will continue opening "I followed README but it doesn't work" issues  when the content conflicts with the latest release, so we have to care about them

AkihiroSuda commented:
On the second thought, a locator plugin should return the YAML contents, not HTTPS URL, so as to support non-HTTP and non-local remotes
